BUGFIX: Ensure that its not attempted to publish 0 events

mhsdesign opened the pull request:
followup to https://github.com/neos/neos-development-collection/pull/5357

see https://github.com/neos/neos-development-collection/pull/5357#issuecomment-2481546116

currently `SetNodeProperties` with 0 properties attempts to publish 0 events which throws an error since #5357

> Writable events must contain at least one event

Instead we throw an exception in `SetNodeProperties`. We DONT expect `PropertyValuesToWrite` to be empty as suggested: https://github.com/neos/neos-development-collection/pull/5357#issuecomment-2481580889

As there is code using `PropertyValuesToWrite` as builder and 

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
BUGFIX: Asset that `SetNodeProperties` is not handled empty

dlubitz reviewed (dismissed):
Looks good by reading.

mhsdesign committed:
TASK: Fix constraint test of failing too early

> InvalidArgumentException (1733394351): The command "SetNodeProperties" for node lady-eleonode-rootford must contain property values

We want a NodeAggregateIsRoot exception

mhsdesign committed:
TASK: Ensure that `Events` are never instantiated empty

by enforcing this via "non-empty-array" we can better catch runtime type errors beforehand via phpstan. As otherwise the event store will fail with

> Writable events must contain at least one event

Some internal utility methods previously also returned possibly empty "Events" instances which were then transformed via iterator_to_array back to arrays and merged. This was not simplified to just work with arrays directly to denote that they might be empty. In cases were we work with fully sure filled intermediate event arrays we use non-empty-array.

With this now fine-grained type checking aside from the original found set properties error following cases will also break with 0 events:

- handleSetSerializedNodeProperties (known)
- handleSetSerializedNodeReferences
- handlePublishWorkspace and handlePublishIndividualNodesFromWorkspace if there were no applied events that implement PublishableToWorkspaceInterface -> goes into direction of https://github.com/neos/neos-development-collection/pull/5337
- structure adjustment TETHERED_NODE_WRONGLY_ORDERED

mhsdesign commented:
Okay before i wanted to merge this i had the feeling we might be missing more such cases ... thats why i attempted to ensure that `Events` are never instantiated empty via https://github.com/neos/neos-development-collection/pull/5396/commits/f0fc051e0365200e0f005ffb4987115400eae2a8

by enforcing this via "non-empty-array" we can better catch runtime type errors beforehand via phpstan. As otherwise the event store will fail with 

> Writable events must contain at least one event

Some internal utility methods previously also returned possibly empty "Events" instances which were then transformed via `iterator_to_array` back to arrays and merged. This was simplified to just work with arrays directly to denote that they might be empty. In cases were fully sure we work with intermediate event arrays wich are annotated via non-empty-array.

With this now fine-grained type checking aside from the original found set properties error following cases will also break with 0 events:

- handleSetSerializedNodeProperties (known)
- handleSetSerializedNodeReferences
- handlePublishWorkspace and handlePublishIndividualNodesFromWorkspace if there were no applied events that implement PublishableToWorkspaceInterface -> goes into direction of https://github.com/neos/neos-development-collection/pull/5337
- structure adjustment TETHERED_NODE_WRONGLY_ORDERED

mhsdesign committed:
BUGFIX: Prevent commit error for structure adjustment TETHERED_NODE_WRONGLY_ORDERED

EventsToPublish must not contain events with no entries

otherwise this error is thrown

> Writable events must contain at least one event

we prevent that by ensuring reorderNodes is called with at least two entries of nodes to be reordered as otherwise there is nothing to do.

mhsdesign committed:
BUGFIX: BUGFIX: Asset that `SetNodeReferences` is not handled empty

mhsdesign committed:
TASK: Add dummy typeguard to prevent phpstan from complaining

it does not understand that the command fields are never empty ... but we ensure that

mhsdesign committed:
TASK: Document why Events does not exist empty

mhsdesign committed:
TASK: Ensure that publishing does not attempt to publish 0 events

The event store would reject that:

> Writable events must contain at least one event

mhsdesign commented:
okay phpstan is happy now and helped me get all the edge cases ... so im happy too ... as this got some new lines changed im requesting a second look before merging this but this seems like a solid sane change

mhsdesign committed:
TASK: Deobscure use of array_keys with TetheredNodeTypeDefinitions::toArray

kitsunet reviewed (approved):
repeating what I said in other places... Willing to go this route but think it's a mistake and we should be strict at commands but lesser so at events if we can (and no ops we def can).

Anywys, this is good +1

mhsdesign commented:
<img width="354" alt="image" src="https://github.com/user-attachments/assets/b8f10579-45f1-46cb-a8a3-5c6ba1e1a94a" />

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php:
```diff
@@ -485,31 +488,32 @@ static function ($handle) use ($commandSimulator, $matchingCommands, $remainingC
-                $workspace->currentContentStreamId
-            );
-            throw $concurrencyException;
+        if ($selectedEventsOfWorkspaceToPublish !== null) {
```
There are no tests describing a possible NULL behaviour.

And the previous comment

> this could empty and a no-op for the rare case when a command returns empty events e.g. the node was already tagged with this subtree tag

Does not apply to 9.0 because we reject such commands.

On the current 9.0 and 9.1 no tests fail if we where to throw an exception here.
